include/envoy/stats/refcount_ptr.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <atomic>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Stats {
+
+// Implements a reference-counted pointer to a class, so that its external usage
+// model is identical to std::shared_ptr, but the reference count itself is held
+// in the class. The class is expected to implement three methods:
+//    void incRefCount()
+//    bool decRefCount()  -- returns true if the reference count goes to zero.
+//    uint32_t use_count()
+// It may implement them by delegating to RefcountHelper (see below), or by
+// inheriting from RefcountInterface (see below).
+template <class T> class RefcountPtr {
+public:
+  RefcountPtr() : ptr_(nullptr) {}
+  RefcountPtr(T* ptr) : ptr_(ptr) {
+    if (ptr_ != nullptr) {
+      ptr_->incRefCount();
+    }
+  }
+
+  RefcountPtr(const RefcountPtr& src) { set(src.get()); }
+
+  // Allows RefcountPtr<BaseClass> foo = RefcountPtr<DerivedClass>.
+  template <class U> RefcountPtr(const RefcountPtr<U>& src) { set(src.get()); }
+
+  // Move-construction is used by absl::flat_hash_map during resizes.
+  RefcountPtr(RefcountPtr&& src) noexcept : ptr_(src.ptr_) { src.ptr_ = nullptr; }
+
+  RefcountPtr& operator=(const RefcountPtr& src) {
+    if (&src != this && src.ptr_ != ptr_) {
+      resetInternal();
+      set(src.get());
+    }
+    return *this;
+  }
+
+  // Move-assignment is used during std::vector resizes.
+  RefcountPtr& operator=(RefcountPtr&& src) noexcept {
+    if (&src != this && src.ptr_ != ptr_) {
+      resetInternal();
+      ptr_ = src.ptr_;
+      src.ptr_ = nullptr;
+    }
+    return *this;
+  }
+
+  ~RefcountPtr() { resetInternal(); }
+
+  // Implements required subset of the shared_ptr API;
+  // see https://en.cppreference.com/w/cpp/memory/shared_ptr for details.
+  T* operator->() const { return ptr_; }
+  T* get() const { return ptr_; }
+  T& operator*() const { return *ptr_; }
+  operator bool() const { return ptr_ != nullptr; }
+  bool operator==(const T* ptr) const { return ptr_ == ptr; }
+  bool operator!=(const T* ptr) const { return ptr_ != ptr; }
+  bool operator==(const RefcountPtr& a) const { return ptr_ == a.ptr_; }
+  bool operator!=(const RefcountPtr& a) const { return ptr_ != a.ptr_; }
+  uint32_t use_count() const { return ptr_->use_count(); }
+  void reset() {
+    if (ptr_ != nullptr) {
+      if (ptr_->decRefCount()) {
+        delete ptr_;
+      }
+      ptr_ = nullptr;
+    }
+  }
+
+private:
+  // Like reset() but does not bother to clear ptr_, as it is about to be
+  // overwritten or destroyed.
+  void resetInternal() {
```

Comment:
maybe make this return a bool for whether it was reset or not, and then just call that from `reset()`?

Reply:
I think I'd prefer to have:
```
void reset() {
  resetInternal();
  ptr_ = nullptr;
}
```
which would be faster than having a conditional.

Reply:
... and done.